fullnode/fullnode_cache.go

```diff
@@ -0,0 +1,228 @@
+
+	// Session cache (5 min on Beta TestNet, see #275)
+	// TODO_MAINNET_MIGRATION(@Olshansk): Revisit after mainnet
+	sessionCache *sturdyc.Client[sessiontypes.Session]
+
+	// The account public key cache; used to cache account public keys indefinitely.
+	// It has an infinite TTL and is populated only once on startup.
+	accountPubKeyCache *sturdyc.Client[cryptotypes.PubKey]
+}
+
+// NewCachingFullNode wraps a LazyFullNode with:
+//   - Session cache: refreshes early to avoid thundering herd/latency spikes
+//   - Account public key cache: indefinite cache for account data
+func NewCachingFullNode(
+	logger polylog.Logger,
+	lazyFullNode *LazyFullNode,
+	cacheConfig CacheConfig,
+) (*cachingFullNode, error) {
+	// Set default session TTL if not set
+	cacheConfig.hydrateDefaults()
+
+	// Log cache configuration
+	logger.Debug().
+		Str("cache_config_session_ttl", cacheConfig.SessionTTL.String()).
+		Msgf("cachingFullNode - Cache Configuration")
+
+	// Configure session cache with early refreshes
+	sessionMinRefreshDelay, sessionMaxRefreshDelay := getCacheDelays(cacheConfig.SessionTTL)
+
+	// Create the session cache with early refreshes
+	sessionCache := sturdyc.New[sessiontypes.Session](
+		cacheCapacity,
+		numShards,
+		cacheConfig.SessionTTL,
+		evictionPercentage,
+		// See: https://github.com/viccon/sturdyc?tab=readme-ov-file#early-refreshes
+		sturdyc.WithEarlyRefreshes(
+			sessionMinRefreshDelay,
+			sessionMaxRefreshDelay,
+			cacheConfig.SessionTTL,
+			retryBaseDelay,
+		),
+	)
+
+	// Create the account cache, which is effectively infinite
+	// caching for the lifetime of the application.
+	// No need to configure early refreshes with SturdyC.
+	accountPubKeyCache := sturdyc.New[cryptotypes.PubKey](
+		cacheCapacity,
+		numShards,
+		pubKeyCacheTTL,
+		evictionPercentage,
+	)
+
+	// Initialize the caching full node with the modified lazy full node
+	return &cachingFullNode{
+		logger:             logger,
+		lazyFullNode:       lazyFullNode,
+		sessionCache:       sessionCache,
+		accountPubKeyCache: accountPubKeyCache,
+	}, nil
+}
+
+// GetApp is a NoOp (apps fetched only at startup; relaying fetches sessions for app/session sync).
+func (cfn *cachingFullNode) GetApp(ctx context.Context, appAddr string) (*apptypes.Application, error) {
+	return nil, fmt.Errorf("GetApp is a NoOp in the caching full node")
+}
+
+// GetSession returns (and auto-refreshes) the session for a service/app from cache.
+func (cfn *cachingFullNode) GetSession(
+	ctx context.Context,
+	serviceID sdk.ServiceID,
+	appAddr string,
+) (sessiontypes.Session, error) {
+	// See: https://github.com/viccon/sturdyc?tab=readme-ov-file#get-or-fetch
+	return cfn.sessionCache.GetOrFetch(
+		ctx,
+		getSessionCacheKey(serviceID, appAddr),
+		func(fetchCtx context.Context) (sessiontypes.Session, error) {
+			cfn.logger.Debug().Str("session_key", getSessionCacheKey(serviceID, appAddr)).Msgf(
+				"[cachingFullNode.GetSession] Making request to full node",
```

Comment:
Rather than addin `"[cachingFullNode.GetSession]` everywhere.

Add a `logger = cfg.logger.With("method", ...)` in the beginning and reuse it.

Ditto everywhere else.